refactor: restructuring function macros + migration notes

benesjan opened the pull request:
In the past core of the macro logic was handled by the external macro code. This has changed and now it's all done by Aztec macro.

This resulted in a lot of the macro functionality now being placed in a way that made no sense. In the PR diff I comment on where each function was moved to.

I also update migration notes and I describe there all the new contract API.

benesjan commented:
* **#18417** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18417?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#18413** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18413?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#18396** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18396?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#18398** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18398?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#18289** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18289?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18289?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#18274** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18274?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#18265** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18265?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17963** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17963?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#18126** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18126?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17792** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17792?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `next`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

benesjan commented on noir-projects/aztec-nr/aztec/src/macros/aztec.nr:
```diff
@@ -315,3 +317,21 @@ comptime fn generate_process_message() -> Quoted {
+
+/// Checks if each function in the module is marked with either #[external(...)], #[contract_library_method], or #[test].
+/// Non-macroified functions are not allowed in contracts.
+comptime fn check_each_fn_macroified(m: Module) {
```
Move here from `noir-projects/aztec-nr/aztec/src/macros/functions/utils.nr`

benesjan commented on noir-projects/aztec-nr/aztec/src/macros/functions/utils.nr:
```diff
@@ -1,527 +0,0 @@
-// TODO(benesjan): Move this whole file to `noir-projects/aztec-nr/aztec/src/macros/internals_functions_generation`.
```
Having this here no longer made any sense. Also having all this functionality in the same file was too messy as these functions are quite complex. All the functionality here has been moved over to `macros/internals_functions_generation/external` directory.

benesjan commented on noir-projects/aztec-nr/aztec/src/macros/functions/utils.nr:
```diff
@@ -1,527 +0,0 @@
-/// Gathers all attributes relevant to the function's ABI and returns a quote that can be applied to the newly generated
-/// function. We apply the abi marker attributes instead of the original ones (e.g. abi_view instead of view) to avoid
-/// the relevant attribute's functionality from getting triggered.
-comptime fn get_abi_relevant_attributes(f: FunctionDefinition) -> Quoted {
```
Moved to `macros/internals_functions_generation/external/helpers.nr`

benesjan commented on noir-projects/aztec-nr/aztec/src/macros/functions/utils.nr:
```diff
@@ -1,527 +0,0 @@
-    attributes
-}
-
-pub(crate) comptime fn transform_private(f: FunctionDefinition) -> Quoted {
```
Moved to `macros/internals_functions_generation/external/private.nr`

benesjan commented on noir-projects/aztec-nr/aztec/src/macros/functions/utils.nr:
```diff
@@ -1,527 +0,0 @@
-    }
-}
-
-pub(crate) comptime fn transform_public(f: FunctionDefinition) -> Quoted {
```
Moved to `macros/internals_functions_generation/external/public.nr`

benesjan commented on noir-projects/aztec-nr/aztec/src/macros/functions/utils.nr:
```diff
@@ -1,527 +0,0 @@
-    }
-}
-
-pub(crate) comptime fn transform_utility(f: FunctionDefinition) -> Quoted {
```
Moved to `macros/internals_functions_generation/external/utility.nr`

benesjan commented on noir-projects/aztec-nr/aztec/src/macros/functions/utils.nr:
```diff
@@ -1,527 +0,0 @@
-
-/// Injects a call to `aztec::messages::discovery::discover_new_messages`, causing for new notes to be added to PXE and made
-/// available for the current execution.
-pub(crate) comptime fn create_message_discovery_call() -> Quoted {
```
Moved to `macros/internals_functions_generation/external/helpers.nr`

benesjan commented on noir-projects/aztec-nr/aztec/src/macros/functions/utils.nr:
```diff
@@ -1,527 +0,0 @@
-/// * `is_private` - Whether the function is a private function (`true`) or a public function (`false`). This determines
-///                  which authwit verification method to use: `assert_current_call_valid_authwit` for private functions
-///                  or `assert_current_call_valid_authwit_public` for public functions.
-pub(crate) comptime fn create_authorize_once_check(
```
Moved to `macros/internals_functions_generation/external/helpers.nr`

benesjan commented on noir-projects/aztec-nr/aztec/src/macros/functions/utils.nr:
```diff
@@ -1,527 +0,0 @@
-
-/// Checks if each function in the module is marked with either #[external(...)], #[contract_library_method], or #[test].
-/// Non-macroified functions are not allowed in contracts.
-pub(crate) comptime fn check_each_fn_macroified(m: Module) {
```
Moved to `macros/aztec.nr` - it was the only place where it was used

benesjan commented on noir-projects/aztec-nr/aztec/src/macros/internals_functions_generation/external/private.nr:
```diff
@@ -0,0 +1,196 @@
+use protocol_types::meta::utils::derive_serialization_quotes;
+use std::meta::{ctstring::AsCtString, type_of};
+
+pub(crate) comptime fn generate_private_external(f: FunctionDefinition) -> Quoted {
```
Renamed the `transform_private` function to this to follow the naming of the internal functions (there it's called `generate_private_internal`.

benesjan commented on noir-projects/aztec-nr/aztec/src/macros/internals_functions_generation/external/public.nr:
```diff
@@ -0,0 +1,144 @@
+};
+use std::meta::ctstring::AsCtString;
+
+pub(crate) comptime fn generate_public_external(f: FunctionDefinition) -> Quoted {
```
Same here. Renamed this.

benesjan commented on noir-projects/aztec-nr/aztec/src/macros/internals_functions_generation/external/utility.nr:
```diff
@@ -0,0 +1,59 @@
+    utils::module_has_storage,
+};
+
+pub(crate) comptime fn generate_utility_external(f: FunctionDefinition) -> Quoted {
```
Same here. Renamed this.

nventuro commented on docs/docs/developers/docs/resources/migration_notes.md:
```diff
@@ -17,9 +17,13 @@ The sandbox command has been renamed and remapped to "local network". We believe
+
+In this release we decided to largely redesign our contract API. Most of the changes here are not a breaking change
+(only renaming of original `#[internal]` to `#[only_self]` and `storage` now being available on the newly introduced
+`self` struct are a breaking change).
```
```suggestion
(only renaming of original `#[internal]` to `#[only_self]` and `context` and `storage` now only being available on the newly introduced
`self` struct are a breaking change).

```

nventuro commented on docs/docs/developers/docs/resources/migration_notes.md:
````diff
@@ -43,9 +47,24 @@ fn _deduct_public_balance(owner: AztecAddress, amount: u64) {
 ```
 
-### Introducing `self` in contracts
+### 2. Introducing of new #[internal]
````
```suggestion
### 2. Introducing new #[internal]
```

Thunkar reviewed (approved):
Lovely

benesjan committed:
refactor: restructuring function macros + migration notes

In the past core of the macro logic was handled by the external macro code. This has changed and now it's all done by Aztec macro.

This resulted in a lot of the macro functionality now being placed in a way that made no sense. In the PR diff I comment on where each function was moved to.

I also update migration notes and I describe there all the new contract API.
